### ISSUE TITLE
Pocket Dimension After re-entering the dimension, there are unloaded pieces of structures.

### ISSUE THREAD

**Issue by EliasLucky** [ConnorTron110/SCP-Lockdown-Issues#37]
If you go out of the pocket dimension into the overworld or any other dimension, and then return back to the pocket dimension. Then you can find chunks with unloaded structures.
If this is not a bug and it was intended to be, please answer me.

Also, when pocket dimension is generating, the world lags accompanying these message in the logs.
`[23:08:58] [Server thread/WARN] [FML]: <mod name> loaded a new chunk [-1, -8] in dimension <dimension id> (<dimension name>) while populating chunk [-2, 14], causing cascading worldgen lag.`

**Comment by ConnorTron110** [comment]
The generation lag is unfortunatly out of my control as it was made by someone who has left the team.

As for the generation bug, I went into the Pocket Dimension (PD) then went to the nether, then to the overworld, then back to the PD and encountered no such bug, if you are able to provide a video or a step by step reproduction that would be great!
Also your Forge Version, and any other mods you may have running will be useful! :)

**Comment by EliasLucky** [comment]
I am sorry for late response, **but I think i found the solution of this problem.**

I decompiled your mod SCP Lockdown (because I was interested how generation of PD was made, and I found a little mistake in code). Structures are being generated in each chunk without offset by +8 blocks by x coordinate and y coordinate, like in minecraft vanilla world generation, without this offset your PD generation causing Cascading Worldgen lag and thats sucks, because this problem is only for 1.12.2 Minecraft version.

Well I did some tests and there is the solution of this problem:
In Utils.java class you should change chunkPosToBlockPosition method to:
```java
public static BlockPos chunkPosToBlockPosition(int chunkX, int chunkZ)
{
	int x = (chunkX << 4) + 8;
	int z = (chunkZ << 4) + 8;
	return new BlockPos(x, 63, z);
}
```

Also if you're interested.
1. My forge version is 1.12.2-14.23.5.2840.
2. Java 8 | jre 1.8.0_333.
3. 16 gb ram.

I hope this will be helpful, because how I understand (you wrote this in mod description) you're working alone on this project. 
Well, have a nice day and goodluck in development SCP Lockdown for 1.16.5!

**Comment by EliasLucky** [comment]
I am sorry for late response, **but I think i found the solution of this problem.**

I decompiled your mod SCP Lockdown (because I was interested how generation of PD was made, and I found a little mistake in code). Structures are being generated in each chunk **without offset by +8 blocks by x coordinate and z coordinate**, like in minecraft vanilla world generation, without this offset your PD generation causing Cascading Worldgen lag and thats sucks, because this problem is only for 1.12.2 Minecraft versions.

**Well the only way to fix this issue is just to make offset +8 blocks by x coordinate and z coordinate.
Just change method chunkPosToBlockPosition in Utils.java to this:**
```java
public static BlockPos chunkPosToBlockPosition(int chunkX, int chunkZ) {
    int x = (chunkX << 4) + 8;
    int z = (chunkZ << 4) + 8;
    return new BlockPos(x, 0, z);
}
```

**Also if you're interested.**
1. My forge version is 1.12.2-14.23.5.2840.
2. Java 8 | jre 1.8.0_333.
3. 16 gb ram.
4. 
I hope this will be helpful, because how I understand (you wrote this in mod description) you're working alone on this project. 
Well, have a nice day and goodluck in development SCP Lockdown for 1.16.5!

**Comment by ConnorTron110** [comment]
I don’t think this is the issue as that was specifically made to target the corner of a chunk. Adding the 8 just centres it, it doesn’t improve performance nor changes anything directly. 

Regardless I finally tested it with and without, and after flying round for a while, I have found almost a line like error in the PD around the coord (-1000, 2768) with the line stretching across east and west. I have a feeling the problem is more complicated than it seems. For 1.16.5 we plan to rewrite the PD anyways. I’ll see what I can do for the next update, and at least try and alleviate the lag experienced in that dimension. However, if it requires me to rewrite the entire generation code, I may sadly just have to leave it as is.
